### PR TITLE
Consolidate build release targets

### PR DESCRIPTION
**Pull request by realgaurav**
Proposing to move more common targets under Makefile.common including build and release targets.
Also updating golang version to _preferred_ version.

### CODE REVIEW

File: Dockerfile.amd64 [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM amd64/golang:1.12.14-alpine3.10
+FROM amd64/golang:1.12.8-alpine3.9
```
Comment (caseydavenport) [inline Dockerfile.amd64]: Could you give some detail on this change? What's the reasoning for this?

File: Makefile.common [hunk 2]
```diff
@@ -353,3 +415,88 @@ help:
...
+ifneq ($(VERSION), $(GIT_VERSION))
+	$(error Attempt to build $(VERSION) from $(GIT_VERSION))
+endif
+
+	$(MAKE) image
+	$(MAKE) tag-images IMAGETAG=$(VERSION)
+
+## Verifies the release artifacts produces by `make release-build` are correct.
+release-verify: release-prereqs
+	# TODO: Check the reported version is correct for each release artifact. Uncomment when binary supports version command.
```
Comment (caseydavenport) [inline Makefile.common]: This is all commented out - how do you propose we re-enable this check? 

One challenge is that each repo has its own version command, so we need a way for the calling Makefile to specify the command to run to get the binary version.

File: Dockerfile.amd64 (hunk 1 above)
Comment (realgaurav) [reply Dockerfile.amd64]: @rafaelvanoni can shed some light. I was asked to use this go-version.

File: Makefile.common [hunk 3]
```diff
@@ -353,3 +415,88 @@ help:
...
+	# fi
+
+## Generates release notes based on commits in this version.
+release-notes: release-prereqs
+	mkdir -p dist
+	echo "# Changelog" > release-notes-$(VERSION)
+	sh -c "git cherry -v $(PREVIOUS_RELEASE) | cut '-d ' -f 2- | sed 's/^/- /' >> release-notes-$(VERSION)"
+
+## Pushes a github release and release artifacts produced by `make release-build`.
+release-publish: release-prereqs
```
Comment (caseydavenport) [inline Makefile.common]: This varies based on repository as well. I don't think this really make sense to have in the common Makefile.

The entire premise of the release infrastructure is that each repository knows how to build and publish itself. Moving it to this file removes the necessary context required to properly do that. 

e.g., the CNI repository publishes binaries to GitHub, but app-policy does not. How do we represent that here?

File: Makefile.common [hunk 4]
```diff
@@ -353,3 +415,88 @@ help:
...
+	@echo ""
+
+# WARNING: Only run this target if this release is the latest stable release. Do NOT
+# run this target for alpha / beta / release candidate builds, or patches to earlier Calico versions.
+## Pushes `latest` release images. WARNING: Only run this for latest stable releases.
+release-publish-latest: release-prereqs
+	$(MAKE) push-all push-non-manifests IMAGETAG=latest
+
+# release-prereqs checks that the environment is configured properly to create a release.
+release-prereqs:
```
Comment (caseydavenport) [inline Makefile.common]: If you look across the core repos I think you'll spot that this target also varies a little bit.

File: Makefile.common [hunk 5]
```diff
@@ -287,6 +287,68 @@ git-push:
...
+	docker tag $(BUILD_IMAGE):latest-$(ARCH) $(call unescapefs,$*:$(IMAGETAG))
+else
+	$(NOECHO) $(NOOP)
+endif
+
+## tag images of all archs
+tag-images-all: imagetag $(addprefix sub-tag-images-,$(VALIDARCHES))
+sub-tag-images-%:
+	$(MAKE) tag-images ARCH=$* IMAGETAG=$(IMAGETAG)
+
```
Comment (doublek) [inline Makefile.common]: @caseydavenport WDYT about just the targets highlighted by this comment move into the common makefile? I agree that the release ones are repo specific and require further discussion.

Comment (caseydavenport) [reply Makefile.common]: I think that's good to do so long as the IMAGETAG / BUILD_IMAGE / PUSH_IMAGES / etc, etc, can be properly set in the main Makefile and interpreted in this one.

**Comment by caseydavenport** [comment]
Closing since this is fairly old and inactive.

**Pull request closed without merging**